

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Project of Escherichia coli &mdash; HUifeng&#39;s BLOG 1.0.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="CurriculumVitae" href="CurriculumVitae.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> HUifeng's BLOG
          

          
          </a>

          
            
            
              <div class="version">
                1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Bio.html">Bio</a></li>
<li class="toctree-l1"><a class="reference internal" href="CurriculumVitae.html">CurriculumVitae</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Project of <em>Escherichia coli</em></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#environment">Environment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genome-download">Genome Download</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#install-wget">Install Wget</a></li>
<li class="toctree-l3"><a class="reference internal" href="#download-via-accession-number">Download via accession number</a></li>
<li class="toctree-l3"><a class="reference internal" href="#translate-into-amino-acid-sequence">Translate into amino acid sequence</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#virulence-factors-detection">Virulence Factors Detection</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#result-analysis">Result Analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#data-visualization">Data visualization</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#use-upsetr">Use UpsetR:</a></li>
<li class="toctree-l4"><a class="reference internal" href="#use-networkx">Use NetworkX</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#in-silico-serotyping">In Silico Serotyping</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#result-statistics">Result Statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#use-prokka-annotate">Use Prokka Annotate</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#build-local-database-of-escherichia-coli">Build local Database of <em>Escherichia coli</em></a></li>
<li class="toctree-l4"><a class="reference internal" href="#run-prokka">Run prokka</a></li>
<li class="toctree-l4"><a class="reference internal" href="#use-roary-get-single-copy-gene">Use Roary get single copy gene</a></li>
<li class="toctree-l4"><a class="reference internal" href="#model-select">Model Select</a></li>
<li class="toctree-l4"><a class="reference internal" href="#use-fasttree-construct-phylogenetic-tree">Use FastTree construct phylogenetic tree</a></li>
<li class="toctree-l4"><a class="reference internal" href="#calculate-population-structure">Calculate Population Structure</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tree-visualization">Tree Visualization</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#visualize-tree-with-pan-genome-profile-use-roary-script">Visualize Tree with Pan-genome Profile( Use Roary script)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualize-tree-with-ggtree">Visualize Tree with ggtree</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">HUifeng's BLOG</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Project of <em>Escherichia coli</em></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/Ecoli.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="project-of-escherichia-coli">
<h1>Project of <em>Escherichia coli</em><a class="headerlink" href="#project-of-escherichia-coli" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Escherichia_coli">Escherichia coli</a> is
a Gram-negative, rod-shaped bacterium belonging to the family
<em>Enterobacteriaceae</em> that was described in 1885 by a German
pediatrician. Pathogenic E.coli is versatile due to the diversity of
their gene sets. Virulence factors usually located on a virulence
plasmid and can be acquired through gene transfer.Different combination
of virulence factors may caused to different illness among human and
animals.</p>
<p>This time, we investigate the co-occurrance of virulence factors among
all the available genome in the
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/">Genbank</a>. Meanwhile, we use
phylogenetic tree to investigate relationship between different O/H
serotype.</p>
</div>
<div class="section" id="environment">
<h2>Environment<a class="headerlink" href="#environment" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference external" href="https://www.python.org/download/releases/2.7/">Python</a> 2.7(
<a class="reference external" href="https://www.anaconda.com/">Anaconda</a>)</li>
</ul>
</div>
<div class="section" id="genome-download">
<h2>Genome Download<a class="headerlink" href="#genome-download" title="Permalink to this headline">¶</a></h2>
<div class="section" id="install-wget">
<h3>Install Wget<a class="headerlink" href="#install-wget" title="Permalink to this headline">¶</a></h3>
<p>On MacOs</p>
<div class="code shell highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ruby</span> <span class="o">-</span><span class="n">e</span> <span class="s2">&quot;$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)&quot;</span>
<span class="n">brew</span> <span class="n">install</span> <span class="n">wget</span>
</pre></div>
</div>
<p>On Windows</p>
<p>Download <a class="reference external" href="http://gnuwin32.sourceforge.net/packages/wget.htm">wget</a></p>
</div>
<div class="section" id="download-via-accession-number">
<h3>Download via accession number<a class="headerlink" href="#download-via-accession-number" title="Permalink to this headline">¶</a></h3>
<p>According the
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt">assembly_summary_genbank</a>
in NCBI ftp site, we can get the genome through the organism name</p>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="s2">&quot;Ecoli_genome&quot;</span><span class="p">)</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="s2">&quot;Ecoli_genome&quot;</span><span class="p">)</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;assembly_summary_genbank.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">2</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;wget &#39;</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="s1">&#39;_genomic.fna.gz&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>or Use <a class="reference external" href="https://pandas.pydata.org/">Pandas</a></p>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">f</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s2">&quot;assembly_summary_genbank.txt&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">header</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f</span><span class="p">[</span><span class="n">f</span><span class="p">[</span><span class="s1">&#39;organism_name&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;Escherichia coli&#39;</span><span class="p">][</span><span class="s1">&#39;ftp_path&#39;</span><span class="p">]:</span>
    <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="s1">&#39;wget &#39;</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="s1">&#39;_genomic.fna.gz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="translate-into-amino-acid-sequence">
<h3>Translate into amino acid sequence<a class="headerlink" href="#translate-into-amino-acid-sequence" title="Permalink to this headline">¶</a></h3>
<p>Usually I only download fna file, and use
<a class="reference external" href="https://github.com/hyattpd/Prodigal">Prodigal</a> to translate
nucleotide into amino acid sequence.</p>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">):</span>
    <span class="n">order</span> <span class="o">=</span> <span class="s1">&#39;prodigal -i &#39;</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="s1">&#39; -q -a &#39;</span><span class="o">+</span><span class="n">i</span><span class="s1">&#39;.faa&#39;</span><span class="o">+</span><span class="s1">&#39; -d &#39;</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="s1">&#39;.nucl&#39;</span><span class="o">+</span><span class="s1">&#39; -o &#39;</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="s1">&#39;.out&#39;</span>
    <span class="nb">print</span> <span class="n">order</span>
    <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">order</span><span class="p">)</span>
</pre></div>
</div>
<p>Now we can get predicted protein sequence of all E.coli genomes.</p>
</div>
</div>
<div class="section" id="virulence-factors-detection">
<h2>Virulence Factors Detection<a class="headerlink" href="#virulence-factors-detection" title="Permalink to this headline">¶</a></h2>
<p>We use all protein sequence to run
<a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> against a In-house
collected gene set:</p>
<div class="code shell highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="o">*.</span><span class="n">faa</span> <span class="o">&gt;</span> <span class="n">Ecoli</span><span class="o">.</span><span class="n">faa</span>
<span class="n">makeblastdb</span> <span class="o">-</span><span class="ow">in</span> <span class="n">VF</span><span class="o">.</span><span class="n">faa</span> <span class="o">-</span><span class="n">dbtype</span> <span class="n">prot</span>
<span class="n">blastp</span> <span class="o">-</span><span class="n">query</span> <span class="n">Ecoli</span><span class="o">.</span><span class="n">faa</span> <span class="o">-</span><span class="n">db</span> <span class="n">VF</span><span class="o">.</span><span class="n">faa</span> <span class="o">-</span><span class="n">out</span> <span class="n">VF</span><span class="o">.</span><span class="n">blast</span> <span class="o">-</span><span class="n">outfmt</span> <span class="mi">6</span> <span class="o">-</span><span class="n">evalue</span> <span class="mf">1e-10</span>  <span class="o">-</span><span class="n">num_threads</span> <span class="mi">8</span> <span class="o">-</span><span class="n">num_alignments</span> <span class="mi">1</span>
</pre></div>
</div>
<div class="section" id="result-analysis">
<h3>Result Analysis<a class="headerlink" href="#result-analysis" title="Permalink to this headline">¶</a></h3>
<p>Now we get the file of
<a class="reference external" href="https://github.com/hzafeng/huifeng/tree/master/source/file">Strain_Profile.csv</a></p>
</div>
<div class="section" id="data-visualization">
<h3>Data visualization<a class="headerlink" href="#data-visualization" title="Permalink to this headline">¶</a></h3>
<div class="section" id="use-upsetr">
<h4>Use <a class="reference external" href="https://cran.r-project.org/web/packages/UpSetR/README.html">UpsetR</a>:<a class="headerlink" href="#use-upsetr" title="Permalink to this headline">¶</a></h4>
<div class="code r highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show coexist between Agg subunit</span>
<span class="n">library</span><span class="p">(</span><span class="n">UpSetR</span><span class="p">)</span>
<span class="n">f</span><span class="o">=</span><span class="n">read</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span><span class="s2">&quot;Strain_Profile.csv&quot;</span><span class="p">)</span>
<span class="n">upset</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">order</span><span class="o">.</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;freq&quot;</span><span class="p">,</span><span class="n">sets</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s1">&#39;Agg3A&#39;</span><span class="p">,</span><span class="s1">&#39;Agg3B&#39;</span><span class="p">,</span><span class="s1">&#39;Agg3C&#39;</span><span class="p">,</span><span class="s1">&#39;Agg3D&#39;</span><span class="p">,</span><span class="s1">&#39;Agg4A&#39;</span><span class="p">,</span><span class="s1">&#39;Agg4B&#39;</span><span class="p">,</span><span class="s1">&#39;Agg4C&#39;</span><span class="p">,</span><span class="s1">&#39;Agg4D&#39;</span><span class="p">,</span><span class="s1">&#39;Agg5A&#39;</span><span class="p">,</span><span class="s1">&#39;AggA&#39;</span><span class="p">,</span><span class="s1">&#39;AggB&#39;</span><span class="p">,</span><span class="s1">&#39;AggC&#39;</span><span class="p">,</span><span class="s1">&#39;AggD&#39;</span><span class="p">))</span>
</pre></div>
</div>
<p><img alt="image0" src="_images/agg_upsetR.png" /></p>
<div class="code r highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show co-exist between Stx toxin</span>
<span class="n">upset</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">order</span><span class="o">.</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;freq&quot;</span><span class="p">,</span><span class="n">sets</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;Stx2aB&quot;</span><span class="p">,</span><span class="s2">&quot;Stx2aA&quot;</span><span class="p">,</span><span class="s2">&quot;Stx2bA&quot;</span><span class="p">,</span><span class="s2">&quot;Stx1aA&quot;</span><span class="p">,</span><span class="s2">&quot;Stx1aB&quot;</span><span class="p">,</span><span class="s2">&quot;Stx2dB&quot;</span><span class="p">,</span><span class="s2">&quot;Stx2bB&quot;</span><span class="p">,</span><span class="s2">&quot;Stx1cA&quot;</span><span class="p">,</span><span class="s2">&quot;Stx1cB&quot;</span><span class="p">,</span><span class="s2">&quot;Stx2dA&quot;</span><span class="p">,</span><span class="s2">&quot;Stx1dB&quot;</span><span class="p">,</span><span class="s2">&quot;Stx2fA&quot;</span><span class="p">,</span><span class="s2">&quot;Stx2fB&quot;</span><span class="p">,</span><span class="s2">&quot;Stx1dA&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p><img alt="image1" src="_images/stx_upsetR.png" /></p>
<div class="code r highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show co-exist between Bfp toxin</span>
<span class="n">upset</span><span class="p">(</span><span class="n">f</span><span class="p">,</span><span class="n">order</span><span class="o">.</span><span class="n">by</span> <span class="o">=</span> <span class="s2">&quot;freq&quot;</span><span class="p">,</span><span class="n">sets</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;BfpA&quot;</span><span class="p">,</span><span class="s2">&quot;BfpB&quot;</span><span class="p">,</span><span class="s2">&quot;BfpC&quot;</span><span class="p">,</span><span class="s2">&quot;BfpD&quot;</span><span class="p">,</span><span class="s2">&quot;BfpE&quot;</span><span class="p">,</span><span class="s2">&quot;BfpF&quot;</span><span class="p">,</span><span class="s2">&quot;BfpG&quot;</span><span class="p">,</span><span class="s2">&quot;BfpH&quot;</span><span class="p">,</span><span class="s2">&quot;BfpI&quot;</span><span class="p">,</span><span class="s2">&quot;BfpJ&quot;</span><span class="p">,</span><span class="s2">&quot;BfpK&quot;</span><span class="p">,</span><span class="s2">&quot;BfpL&quot;</span><span class="p">,</span><span class="s2">&quot;BfpP&quot;</span><span class="p">,</span><span class="s2">&quot;BfpU&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p><img alt="image2" src="_images/bfp_upsetR.png" /></p>
</div>
<div class="section" id="use-networkx">
<h4>Use <a class="reference external" href="https://networkx.github.io/documentation/latest/index.html">NetworkX</a><a class="headerlink" href="#use-networkx" title="Permalink to this headline">¶</a></h4>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show co-exist between Shiga toxin and [Simplified Adhesin Factors]</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="k">def</span> <span class="nf">get_co_oc_from_show_vfs</span><span class="p">(</span><span class="n">show_vf</span><span class="p">):</span>
    <span class="n">cooc</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;adhesin_simplify.csv&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">data</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
            <span class="n">show_vf</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">show_vf</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">show_vf</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">j</span><span class="o">&gt;</span><span class="n">i</span><span class="p">:</span>
                <span class="n">con</span><span class="o">=</span><span class="mi">0</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">strain_profile</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">show_vf</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="ow">in</span> <span class="n">strain_profile</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="ow">and</span> <span class="n">show_vf</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="ow">in</span> <span class="n">strain_profile</span><span class="p">[</span><span class="n">k</span><span class="p">]:</span>
                        <span class="n">con</span><span class="o">+=</span><span class="mi">1</span>
                <span class="k">if</span> <span class="n">con</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">cooc</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">show_vf</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="n">show_vf</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">con</span><span class="p">))</span>
<span class="n">show_vf</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Stx1aA&#39;</span><span class="p">,</span><span class="s1">&#39;Stx1cA&#39;</span><span class="p">,</span><span class="s1">&#39;Stx1dA&#39;</span><span class="p">,</span><span class="s1">&#39;Stx2aA&#39;</span><span class="p">,</span><span class="s1">&#39;Stx2bA&#39;</span><span class="p">,</span><span class="s1">&#39;Stx2dA&#39;</span><span class="p">,</span><span class="s1">&#39;Stx2eA&#39;</span><span class="p">,</span><span class="s1">&#39;Stx1aB&#39;</span><span class="p">,</span><span class="s1">&#39;Stx1cB&#39;</span><span class="p">,</span><span class="s1">&#39;Stx1dB&#39;</span><span class="p">,</span><span class="s1">&#39;Stx2aB&#39;</span><span class="p">,</span><span class="s1">&#39;Stx2bB&#39;</span><span class="p">,</span><span class="s1">&#39;Stx2dB&#39;</span><span class="p">,</span><span class="s1">&#39;Stx2eB&#39;</span><span class="p">]</span>
<span class="n">get_co_oc_from_show_vfs</span><span class="p">(</span><span class="n">show_vf</span><span class="p">)</span>
<span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">()</span>
<span class="n">G</span><span class="o">=</span><span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
<span class="n">G</span><span class="o">.</span><span class="n">add_weighted_edges_from</span><span class="p">(</span><span class="n">cooc</span><span class="p">)</span>
<span class="n">weights</span> <span class="o">=</span><span class="p">[]</span>
<span class="n">weights</span><span class="o">=</span> <span class="p">[</span><span class="n">w</span><span class="p">[</span><span class="mi">2</span><span class="p">][</span><span class="s1">&#39;weight&#39;</span><span class="p">]</span><span class="o">*</span><span class="mf">0.001</span> <span class="k">for</span> <span class="n">w</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="kc">True</span><span class="p">)]</span>
<span class="n">node_color</span><span class="o">=</span><span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
    <span class="k">if</span> <span class="s1">&#39;Stx&#39;</span> <span class="ow">in</span> <span class="n">i</span><span class="p">:</span>
        <span class="n">node_color</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;#CD3333&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">node_color</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;#B5B5B5&#39;</span><span class="p">)</span>
<span class="n">pos</span><span class="o">=</span><span class="n">nx</span><span class="o">.</span><span class="n">circular_layout</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
<span class="n">nx</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">G</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span><span class="n">node_size</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="n">weights</span><span class="p">,</span><span class="n">with_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">font_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span><span class="n">node_color</span><span class="o">=</span><span class="n">node_color</span><span class="p">,</span><span class="n">edge_color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
</pre></div>
</div>
<p><img alt="image3" src="_images/Stx_ad_circular.png" /></p>
<p>Adhesin factors and heat stable toxin</p>
<p><img alt="image4" src="_images/St_ad_circular.png" /></p>
<p>Adhesin factors and heat labile toxin</p>
<p><img alt="image5" src="_images/Elt_ad_circular.png" /></p>
<p>All toxin</p>
<p><img alt="image6" src="_images/toxin_circular.png" /></p>
<p><a class="reference external" href="https://github.com/hzafeng/huifeng/tree/master/source/images/PDF">Get PDF
File</a></p>
</div>
</div>
</div>
<div class="section" id="in-silico-serotyping">
<h2>In Silico Serotyping<a class="headerlink" href="#in-silico-serotyping" title="Permalink to this headline">¶</a></h2>
<p>To understand the relationship between different serotype and their
virulence profile, we use
<a class="reference external" href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> based in silico
methods to serotype <em>E.coli</em> via
<a class="reference external" href="https://cge.cbs.dtu.dk/services/SerotypeFinder/">SerotypeFinder</a>.</p>
<div class="code shell highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="o">*.</span><span class="n">faa</span> <span class="o">&gt;</span> <span class="n">Ecoli</span><span class="o">.</span><span class="n">faa</span>
<span class="n">makeblastdb</span> <span class="o">-</span><span class="ow">in</span> <span class="n">serotype</span><span class="o">.</span><span class="n">faa</span> <span class="o">-</span><span class="n">dbtype</span> <span class="n">prot</span>
<span class="n">blastp</span> <span class="o">-</span><span class="n">query</span> <span class="n">Ecoli</span><span class="o">.</span><span class="n">faa</span> <span class="o">-</span><span class="n">db</span> <span class="n">serotype</span><span class="o">.</span><span class="n">faa</span> <span class="o">-</span><span class="n">out</span> <span class="n">EcoliSerotype</span><span class="o">.</span><span class="n">blast</span> <span class="o">-</span><span class="n">outfmt</span> <span class="mi">6</span> <span class="o">-</span><span class="n">evalue</span> <span class="mf">1e-10</span>  <span class="o">-</span><span class="n">num_threads</span> <span class="mi">8</span> <span class="o">-</span><span class="n">num_alignments</span> <span class="mi">1</span>
</pre></div>
</div>
<div class="section" id="result-statistics">
<h3>Result Statistics<a class="headerlink" href="#result-statistics" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>10450 isolate with O antigen</li>
<li>11657 isolate with H antigen</li>
<li>10361 isolate with O&amp;H antigen(including 1576 serotype)</li>
<li>Use O&amp;H antigen and only O antigen for phylogenetic analysis(totally
1616 isolate)</li>
</ul>
<p>Use
<a class="reference external" href="https://github.com/sanger-pathogens/assembly-stats">Assembly-stat</a>
to get assembly quality of each genome, and each serotype group select
top genome completeness isolate to construct phylogeneitc tree. ##
Phylogenetic Tree Construction</p>
</div>
<div class="section" id="use-prokka-annotate">
<h3>Use <a class="reference external" href="https://github.com/tseemann/prokka">Prokka</a> Annotate<a class="headerlink" href="#use-prokka-annotate" title="Permalink to this headline">¶</a></h3>
<div class="section" id="build-local-database-of-escherichia-coli">
<h4>Build local Database of <em>Escherichia coli</em><a class="headerlink" href="#build-local-database-of-escherichia-coli" title="Permalink to this headline">¶</a></h4>
<p>Use all “Complete genome” level of genome completeness strain to
construct database. Download gbk file same as <a class="reference external" href="#download-via-accession-number">Download via accession
number</a>
(<a class="reference external" href="https://huifeng.readthedocs.io/en/latest/Ecoli.html#download-via-accession-number">https://huifeng.readthedocs.io/en/latest/Ecoli.html#download-via-accession-number</a>)</p>
<div class="code shell highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">prokka</span><span class="o">-</span><span class="n">genbank_to_fasta_db</span> <span class="o">*.</span><span class="n">gbk</span> <span class="o">&gt;</span> <span class="n">ecoli</span><span class="o">.</span><span class="n">faa</span>
<span class="n">cd</span><span class="o">-</span><span class="n">hit</span> <span class="o">-</span><span class="n">i</span> <span class="n">ecoli</span><span class="o">.</span><span class="n">faa</span> <span class="o">-</span><span class="n">o</span> <span class="n">ecoli</span> <span class="o">-</span><span class="n">T</span> <span class="mi">0</span> <span class="o">-</span><span class="n">M</span> <span class="mi">0</span> <span class="o">-</span><span class="n">g</span> <span class="mi">1</span> <span class="o">-</span><span class="n">s</span> <span class="mf">0.8</span> <span class="o">-</span><span class="n">c</span> <span class="mf">0.9</span>
<span class="n">rm</span> <span class="o">-</span><span class="n">fv</span> <span class="n">ecoli</span><span class="o">.</span><span class="n">faa</span> <span class="n">ecoli</span><span class="o">.</span><span class="n">bak</span><span class="o">.</span><span class="n">clstr</span> <span class="n">ecoli</span><span class="o">.</span><span class="n">clstr</span>
<span class="n">makeblastdb</span> <span class="o">-</span><span class="n">dbtype</span> <span class="n">prot</span> <span class="o">-</span><span class="ow">in</span> <span class="n">ecoli</span>
<span class="n">mv</span> <span class="n">ecoli</span><span class="o">.</span><span class="n">p</span><span class="o">*</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">prokka</span><span class="o">/</span><span class="n">db</span><span class="o">/</span><span class="n">genus</span><span class="o">/</span>
</pre></div>
</div>
</div>
<div class="section" id="run-prokka">
<h4>Run prokka<a class="headerlink" href="#run-prokka" title="Permalink to this headline">¶</a></h4>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">):</span>
    <span class="n">prokka</span> <span class="o">=</span> <span class="s1">&#39;prokka --usegenus --genus ecoli --outdir ./prokka_out&#39;</span><span class="o">+</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39; --locustag &#39;</span><span class="o">+</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39; --prefix &#39;</span><span class="o">+</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39; --cpus 32 &#39;</span><span class="o">+</span><span class="n">i</span>
    <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">prokka</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="use-roary-get-single-copy-gene">
<h4>Use <a class="reference external" href="https://github.com/sanger-pathogens/Roary">Roary</a> get single copy gene<a class="headerlink" href="#use-roary-get-single-copy-gene" title="Permalink to this headline">¶</a></h4>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">roary</span> <span class="o">-</span><span class="n">a</span>  <span class="c1"># Check dependency</span>
<span class="n">mkdir</span> <span class="n">gff</span><span class="p">;</span><span class="n">mkdir</span> <span class="n">ffn</span><span class="p">;</span> <span class="n">mkdir</span> <span class="n">faa</span>
<span class="n">mv</span> <span class="o">*.</span><span class="n">gff</span> <span class="o">./</span><span class="n">gff</span><span class="p">;</span><span class="n">mv</span> <span class="o">*.</span><span class="n">ffn</span> <span class="o">./</span><span class="n">ffn</span><span class="p">;</span> <span class="n">mv</span> <span class="o">*.</span><span class="n">faa</span> <span class="o">./</span><span class="n">faa</span>
<span class="n">cat</span> <span class="o">./</span><span class="n">ffn</span><span class="o">/*.</span><span class="n">ffn</span> <span class="o">&gt;</span> <span class="o">./</span><span class="n">ecoli</span><span class="o">.</span><span class="n">ffn</span>
<span class="n">raory</span> <span class="o">*.</span><span class="n">gff</span>
<span class="c1"># get single copy gene</span>
</pre></div>
</div>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">f</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;gene_presence_absence.csv&quot;</span><span class="p">,</span><span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">f1</span><span class="o">=</span><span class="n">f</span><span class="p">[(</span><span class="n">f</span><span class="p">[</span><span class="s1">&#39;No. isolates&#39;</span><span class="p">]</span><span class="o">==</span><span class="mi">1616</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">f</span><span class="p">[</span><span class="s1">&#39;Avg sequences per isolate&#39;</span><span class="p">]</span><span class="o">==</span><span class="mf">1.0</span><span class="p">]</span>
<span class="n">strain_id</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f1</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;GCA_&#39;</span> <span class="ow">in</span> <span class="n">i</span><span class="p">]</span>
<span class="n">strains</span><span class="o">=</span><span class="n">f1</span><span class="p">[</span><span class="n">strain_id</span><span class="p">]</span>
<span class="n">strain_list</span><span class="o">=</span><span class="n">strains</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">id2seq</span><span class="o">=</span><span class="p">{}</span>
<span class="c1">## Get a dict of id --&gt; seq</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;ecoli.ffn&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">data</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
        <span class="n">id2seq</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span><span class="o">=</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="c1">## write each family sequemces</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">68</span><span class="p">):</span>
    <span class="n">of_name</span><span class="o">=</span><span class="s1">&#39;family&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="n">of</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="n">of_name</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">strain_list</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
        <span class="n">of</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="o">+</span><span class="n">j</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">+</span><span class="n">d</span><span class="p">[</span><span class="n">j</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">of</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>Use <a class="reference external" href="https://www.drive5.com/muscle/">Muscle</a> to align sequence</p>
<div class="code shell highlight-default notranslate"><div class="highlight"><pre><span></span>for i in family*
do
    muscle -in $i -out $i+&#39;_aligned.fasta&#39;
done
</pre></div>
</div>
<p>Connect each alignment file:</p>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="n">dicts</span><span class="o">=</span><span class="p">{}</span> <span class="c1">## Use strainID as key, add each sequence in to the value</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">):</span>
    <span class="k">if</span> <span class="s1">&#39;_aligned.fasta&#39;</span> <span class="ow">in</span> <span class="n">i</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">data</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">j</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">dicts</span><span class="p">:</span>
                    <span class="n">dicts</span><span class="p">[</span><span class="n">j</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span><span class="o">=</span><span class="n">j</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">dicts</span><span class="p">[</span><span class="n">j</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span><span class="o">+=</span><span class="n">j</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">outfile</span><span class="o">=</span><span class="nb">open</span><span class="p">(</span><span class="s1">&#39;connected.faa&#39;</span><span class="p">,</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">dicts</span><span class="p">:</span>
    <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="o">+</span><span class="n">i</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">+</span><span class="n">dicts</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">outfile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>In this way, we can get connected.fna file of connected DNA sequence for
Population Structure calculation</p>
</div>
<div class="section" id="model-select">
<h4>Model Select<a class="headerlink" href="#model-select" title="Permalink to this headline">¶</a></h4>
<p>Use <a class="reference external" href="https://github.com/ddarriba/prottest3">ProtTest</a></p>
<div class="code shell highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">java</span> <span class="o">-</span><span class="n">jar</span> <span class="n">prottest</span><span class="o">-</span><span class="mf">3.4</span><span class="o">.</span><span class="mf">2.</span><span class="n">jar</span> <span class="o">-</span><span class="n">i</span> <span class="n">connected</span><span class="o">.</span><span class="n">faa</span> <span class="o">-</span><span class="n">S</span> <span class="mi">2</span> <span class="o">-</span><span class="nb">all</span><span class="o">-</span><span class="n">distributions</span> <span class="o">-</span><span class="nb">all</span> <span class="o">-</span><span class="n">tc</span> <span class="mf">0.5</span> <span class="o">&gt;</span> <span class="n">prottest</span><span class="o">.</span><span class="n">output</span>
</pre></div>
</div>
<p>In this work, we get “JTT+I+G” model.</p>
</div>
<div class="section" id="use-fasttree-construct-phylogenetic-tree">
<h4>Use <a class="reference external" href="http://www.microbesonline.org/fasttree/">FastTree</a> construct phylogenetic tree<a class="headerlink" href="#use-fasttree-construct-phylogenetic-tree" title="Permalink to this headline">¶</a></h4>
<div class="code shell highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">FastTree</span> <span class="o">-</span><span class="n">gamma</span> <span class="n">connected</span><span class="o">.</span><span class="n">faa</span> <span class="o">&gt;</span> <span class="n">fast</span><span class="o">.</span><span class="n">tree</span>
</pre></div>
</div>
</div>
<div class="section" id="calculate-population-structure">
<h4>Calculate Population Structure<a class="headerlink" href="#calculate-population-structure" title="Permalink to this headline">¶</a></h4>
<p>Use <a class="reference external" href="https://github.com/gtonkinhill/rhierbaps">rhierbaps</a>(R
implementation of
<a class="reference external" href="http://www.helsinki.fi/bsg/software/BAPS/">hierBAPS</a>)</p>
<div class="code r highlight-default notranslate"><div class="highlight"><pre><span></span>devtools::install_github(&quot;gtonkinhill/rhierbaps&quot;)
library(rhierbaps)
snp.matrix &lt;- load_fasta(connected.fasta)
hb.results &lt;- hierBAPS(snp.matrix, max.depth = 2, n.pops = 12, quiet = TRUE)
write.csv(&quot;BAPS.csv&quot;,hb.results$partition.df)
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="tree-visualization">
<h2>Tree Visualization<a class="headerlink" href="#tree-visualization" title="Permalink to this headline">¶</a></h2>
<div class="section" id="visualize-tree-with-pan-genome-profile-use-roary-script">
<h3>Visualize Tree with Pan-genome Profile( Use <a class="reference external" href="https://github.com/sanger-pathogens/Roary/tree/master/contrib/roary_plots">Roary script</a>)<a class="headerlink" href="#visualize-tree-with-pan-genome-profile-use-roary-script" title="Permalink to this headline">¶</a></h3>
<div class="code shell highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">roary_plots</span><span class="o">.</span><span class="n">py</span> <span class="n">fast</span><span class="o">.</span><span class="n">tree</span> <span class="n">gene_presence_absence</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p><img alt="image7" src="_images/pangenome_matrix.png" /></p>
</div>
<div class="section" id="visualize-tree-with-ggtree">
<h3>Visualize Tree with <a class="reference external" href="https://github.com/GuangchuangYu/ggtree">ggtree</a><a class="headerlink" href="#visualize-tree-with-ggtree" title="Permalink to this headline">¶</a></h3>
<div class="code r highlight-default notranslate"><div class="highlight"><pre><span></span>library(ggtree)
library(ggplot2)
cluster=read.csv(&quot;BAPS_reid.csv&quot;)
info&lt;-read.csv(&quot;label2.csv&quot;)
heatmapData&lt;-read.csv(&quot;show_vf_profile.csv&quot;,row.names=2)
heatmapData=heatmapData[2:12]
rn &lt;- rownames(heatmapData)
heatmapData &lt;- as.data.frame(sapply(heatmapData, as.character))
rownames(heatmapData) &lt;- rn

f&lt;-read.tree(&quot;Out_root.nwk.txt&quot;)
group=split(cluster$isolate,cluster$group_level_1)
f &lt;- groupOTU(f, group)
p&lt;-ggtree(f,aes(color=group)) %&lt;+% info +
geom_tiplab(aes(label=Serotype),align=T, linetype=NA,size=0.8)+theme(legend.position = &quot;right&quot;)+scale_color_manual(values=c(&#39;black&#39;,&#39;#8DD3C7&#39;,&#39;#FFFFB3&#39;,&#39;#BEBADA&#39;,&#39;#FB8072&#39;,&#39;#80B1D3&#39;,&#39;#FDB462&#39;,&#39;#B3DE69&#39;,&#39;#FCCDE5&#39;, &#39;#D9D9D9&#39; ,&#39;#BC80BD&#39;, &#39;#CCEBC5&#39;, &#39;#FFED6F&#39;))

p3&lt;-gheatmap(p,heatmapData,offset=0.001,colnames_position=&quot;top&quot;,
             colnames_angle=90,color=NULL,
             colnames_offset_y = 1,hjust=-0.1,
             font.size=1.2)+
scale_fill_manual(values = c(&quot;white&quot;,&quot;Red&quot;),breaks=c(&quot;0&quot;,&quot;1&quot;))
</pre></div>
</div>
<p><img alt="image8" src="_images/ggtree_heatmap_normal.png" /> <img alt="image9" src="_images/ggtree_heatmap_circular.png" /></p>
<p>Get
<a class="reference external" href="https://github.com/hzafeng/huifeng/tree/master/source/images/PDF">PDF</a></p>
<p><em>Shigella Piasmid Mxi-Spa identifed if MxiM or Spa33 existed.</em></p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="CurriculumVitae.html" class="btn btn-neutral" title="CurriculumVitae" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, huifeng@ualberta.ca.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>